### PR TITLE
HPCC-21295 Create TargetClusters Widget

### PR DESCRIPTION
**Pull request by miguelvazq**
Fixes: HPCC-21295

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by miguelvazq** [comment]
@GordonSmith Please review. I am adding each component at a time but, there are services in ESPPreflight.ts that all components need.

File: esp/src/eclwatch/TargetClustersQueryWidget.js [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+                        sortable: false,
+                        width: 75
+                    },
+                    Directory: {
+                        label: this.i18n.Directory,
+                        sortable: false,
+                        width: 200
+                    },
+                }
+            }, domID);
```
Comment (GordonSmith) [inline esp/src/eclwatch/TargetClustersQueryWidget.js]: Should 1 col have no width?

File: esp/src/src/ESPPreflight.ts [hunk 2]
```diff
@@ -1,10 +1,260 @@
...
+    getOS: function (int) {
+        switch (int) {
+            case 0:
+                return "Windows"
+            case 1:
+                return "Solaris"
+            case 2:
+                return "Linux"
+            default:
+                return "Linux"
```
Comment (GordonSmith) [inline esp/src/src/ESPPreflight.ts]: Minor:  Shorten to:
```
            case 2:
            default:
                return "Linux"
```

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/eclwatch/MachineInformationWidget.js [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    "src/Utility",
+    "src/ws_machine",
+
+    "dojo/text!../templates/MachineInformationWidget.html",
+
+    "dijit/form/DropDownButton",
+    "dijit/TooltipDialog",
+    "dijit/form/Form",
+    "dijit/form/Button",
+    "dijit/form/NumberTextBox",
```
[the remark this answers is not in the document]
Comment (miguelvazq) [reply esp/src/eclwatch/MachineInformationWidget.js]: TargetClustersQueryWidget seems to open in new page just fine. MachineInformationWidget is just a filter dialog so there is no open in new window functionality.

File: esp/src/eclwatch/TargetClustersQueryWidget.js (hunk 1 above)
Comment (miguelvazq) [reply esp/src/eclwatch/TargetClustersQueryWidget.js]: When applying no width to a column in this grid it causes all sorts of column sizing issues.

**Comment by miguelvazq** [comment]
@GordonSmith Changes made.

File: esp/src/eclwatch/LogWidget.js [hunk 4]
```diff
@@ -142,18 +142,30 @@ define([
 
         //  Implementation  ---
         init: function (params) {
-            if (this.params.__hpcc_id === params.__hpcc_id)
+
+            this.logTargetSelect.reset();
+            if (params.newPreflight) {
+                if (this.inherited(arguments))
+                    return;
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply esp/src/eclwatch/LogWidget.js]: @miguelvazq bump

File: esp/src/eclwatch/MachineInformationWidget.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (GordonSmith) [reply esp/src/eclwatch/MachineInformationWidget.js]: Ok

File: esp/src/src/ESPPreflight.ts (hunk 2 above)
Comment (GordonSmith) [reply esp/src/src/ESPPreflight.ts]: Minor:  But if your making changes for the inherited init call, you may as well do this as well.

File: esp/src/eclwatch/LogWidget.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (miguelvazq) [reply esp/src/eclwatch/LogWidget.js]: It is now outside the `if (params.newPreflight)`

**Review by GordonSmith — APPROVED** [review]
Looks good.

**Pull request merged**